File: backend/src/Services/UserService.cs
using API.Entities;
using API.Interfaces.Repositories;
using API.Types;
using API.Utils;

namespace API.Services;

public class UserService(IUserRepository userRepository)
{

    public async Task<Page<User>> GetAll(string filter, int page, int limit) {

        Page<User> users = await userRepository.GetAll(filter, page, limit);

        return users;

    }

    public async Task<User?> GetUserById(int id) {

        User? user = await userRepository.GetById(id);

        return user;

    }

    public async Task<User?> GetUserByUsername(string username) {

        User? user = await userRepository.GetByUsername(username);

        return user;

    }

    public async Task<User?> CreateUser(string username, string password, string fullName, string email, string phoneNumber, Role role) {

        User user = new() {
            Username = username,
            Password = BCrypt.Net.BCrypt.HashPassword(password, 10),
            FullName = fullName,
            Email = email,
            PhoneNumber = phoneNumber,
            Role = role,
            JTI = JsonWebTokenUtils.GenerateJTI(5)
        };

        try {
            return await userRepository.Create(user);
        } catch {
            return null;
        }

    }

    public async Task<User?> UpdateUser(int id, string? username, string? password, string? fullName, string? email, string? phoneNumber, Role? role) {

        User? user = await userRepository.GetById(id);

        if(user == null) {
            return null;
        }

        user.Username = username ?? user.Username;

        if(password != null) {
            user.Password = BCrypt.Net.BCrypt.HashPassword(password, 10);
            user.JTI = JsonWebTokenUtils.GenerateJTI(5);
        }

        user.FullName = fullName ?? user.FullName;
        user.Email = email ?? user.Email;
        user.PhoneNumber = phoneNumber ?? user.PhoneNumber;
        user.Role = role ?? user.Role;

        try {
            return await userRepository.Update(user);
        } catch {
            return null;
        }

    }

    public async Task<User?> DeleteUser(int id) {

        User? user = await userRepository.GetById(id);

        if(user == null) {
            return null;
        }

        try {
            return await userRepository.Delete(user);
        } catch {
            return null;
        }

    }

}

File: frontend/src/services/user.service.ts
import { APIError } from "@/models/api-error.models";
import { Page } from "@/models/page";
import { User } from "@/models/user.model";
import { Service } from "@/services/service";
import { optional } from "@/types/optional";
import { Role } from "@/types/role.enum";

export class UserService extends Service {

    static async getUsers(filter : string, page : number, limit : number) {

        const response = await this.axios.get<Page<User> | APIError>('/api/user', {
            params: {
                filter,
                page,
                limit
            }
        });

        return response.data;

    }

    static async createUser(username : string, password : string, fullName : string, email : string, phoneNumber : string, role : Role) {

        const response = await this.axios.post<User | APIError>('/api/user', {
            username,
            password,
            fullName,
            email,
            phoneNumber,
            role
        });

        return response.data;

    }

    static async updateUser(id : number, username : optional<string>, password : optional<string>, fullName : optional<string>, email : optional<string>, phoneNumber : optional<string>, role : optional<Role>) {

        const response = await this.axios.put<User | APIError>(`/api/user/${id}`, {
            username,
            password,
            fullName,
            email,
            phoneNumber,
            role
        });

        return response.data;

    }

    static async deleteUser(id : number) {

        const response = await this.axios.delete<User | APIError>(`/api/user/${id}`);

        return response.data;

    }

}